### PR TITLE
Natasha/series detial and subscriptions backend

### PR DESCRIPTION
**Pull request by natashaarmbrust**
massive PR sorry about that

### CODE REVIEW

File: Podcast/EndpointRequest.swift [hunk 1]
```diff
@@ -101,7 +101,7 @@ class EndpointRequest: Operation {
         return baseURLString + path
     }
     
-    func parameters() -> [String:Any] {
+    func parameters() -> [String:Any]? {
```
Comment (dennisfedorko) [inline Podcast/EndpointRequest.swift]: why did you change this?

File: Podcast/EndpointRequest.swift [hunk 2]
```diff
@@ -113,6 +113,10 @@ class EndpointRequest: Operation {
             return localQueryParameters
         }
         
+        if params.count == 0 {
```
Comment (dennisfedorko) [inline Podcast/EndpointRequest.swift]: why do you return nil instead of an empty dict?

File: Podcast/Episode.swift [hunk 3]
```diff
@@ -13,26 +13,26 @@ class Episode: NSObject {
     
     var id: Int
     var title: String
-    var series: Series?
+    var seriesId: Int
```
Comment (dennisfedorko) [inline Podcast/Episode.swift:16]: seriesID: String

File: Podcast/Episode.swift [hunk 4]
```diff
@@ -58,29 +58,47 @@ class Episode: NSObject {
...
      }
     
     // Returns data - time - series in a string
     func dateTimeSeriesString() -> String {
         let dateFormatter = DateFormatter()
         dateFormatter.dateStyle = .long
         dateFormatter.timeStyle = .none
+        let length = duration
+        
+        //add back in later when we figure out what the data is consistently
```
Comment (dennisfedorko) [inline Podcast/Episode.swift:85]: can you resolve this before you merge

File: Podcast/EpisodesBySeriesIdEndpointRequest.swift [hunk 5]
```diff
@@ -0,0 +1,31 @@
...
+//  Copyright © 2017 Cornell App Development. All rights reserved.
+//
+
+import UIKit
+import SwiftyJSON
+
+class EpisodesBySeriesIdEndpointRequest: EndpointRequest {
+    
+    var seriesID: String
+    var offset: Int!
```
Comment (dennisfedorko) [inline Podcast/EpisodesBySeriesIdEndpointRequest.swift]: doesn't need to be Int! since you initialize in init, same with max, they can both be Int

File: Podcast/EpisodesBySeriesIdEndpointRequest.swift [hunk 6]
```diff
@@ -0,0 +1,31 @@
...
+//  Podcast
+//
+//  Created by Natasha Armbrust on 3/30/17.
+//  Copyright © 2017 Cornell App Development. All rights reserved.
+//
+
+import UIKit
+import SwiftyJSON
+
+class EpisodesBySeriesIdEndpointRequest: EndpointRequest {
```
Comment (dennisfedorko) [inline Podcast/EpisodesBySeriesIdEndpointRequest.swift]: maybe FetchEpisodesForSeriesIDEndpointRequest is a better name?

File: Podcast/EpisodesBySeriesIdEndpointRequest.swift [hunk 7]
```diff
@@ -0,0 +1,31 @@
...
+        self.offset = offset
+        self.max = max
+        super.init()
+        path = "/podcasts/episodes/" + seriesID
+        httpMethod = .get
+        queryParameters = ["max": max, "offset": offset]
+    }
+    
+    override func processResponseJSON(_ json: JSON) {
+        processedResponseValue = json["data"]["episodes"].map{ series in Episode(json: series.1) }
```
Comment (dennisfedorko) [inline Podcast/EpisodesBySeriesIdEndpointRequest.swift]: Im not a fan of this map format, it's not possible to understand the type of series from reading this. Also why do you use series.1?

File: Podcast/SearchSeriesTableViewCell.swift [hunk 8]
```diff
@@ -81,15 +81,28 @@ class SearchSeriesTableViewCell: UITableViewCell {
...
     }
     
     func didPressSubscribeButton() {
-        subscribeButtonPressed = !subscribeButtonPressed
-        subscribeButton.isSelected = subscribeButtonPressed
-        delegate?.searchSeriesTableViewCell(cell: self, didSetSubscribeButton: subscribeButtonPressed)
+        delegate?.searchSeriesTableViewCellDidPressSubscribeButton(cell: self)
+    }
+    
+    func setSubscribeButtonToState(state: Bool) {
```
Comment (dennisfedorko) [inline Podcast/SearchSeriesTableViewCell.swift]: instead of state, call the parameter isSelected: Bool. makes more sense when reading if isSelected {...}

File: Podcast/SearchTableViewController.swift [hunk 9]
```diff
@@ -7,6 +7,7 @@
 //
 
 import UIKit
+import NVActivityIndicatorView
```
Comment (dennisfedorko) [inline Podcast/SearchTableViewController.swift:10]: is this the activity indicator view we're going to be using everywhere? if so I think we should make an extension on UIViewController to show the indicator view (kind of like I did with the action sheet) so you don't have to import and configure the activity indicator in every VC

File: Podcast/SearchTableViewController.swift [hunk 10]
```diff
@@ -157,4 +160,32 @@ class SearchTableViewController: UITableViewController, SearchEpisodeTableViewCe
         appDelegate.showPlayer(animated: true)
         Player.sharedInstance.playEpisode(episode: episode)
     }
+    
+    func searchSeriesTableViewCellDidPressSubscribeButton(cell: SearchSeriesTableViewCell) {
+        guard let indexPath = tableView.indexPath(for:cell), let series = searchResults[.series]?[indexPath.row] as? Series else { return }
+        series.isSubscribed = !series.isSubscribed
+        if series.isSubscribed { //subscribing to series
```
Comment (dennisfedorko) [inline Podcast/SearchTableViewController.swift]: I dont think your comment is adding much here

File: Podcast/SeriesDetailHeaderView.swift [hunk 11]
```diff
@@ -7,9 +7,10 @@
 //
 
 import UIKit
+import NVActivityIndicatorView
```
Comment (dennisfedorko) [inline Podcast/SeriesDetailHeaderView.swift]: maybe we should make it so only controllers can create activity indicators?

File: Podcast/SeriesDetailViewController.swift [hunk 12]
```diff
@@ -20,76 +21,139 @@ class SeriesDetailViewController: UIViewController, SeriesDetailHeaderViewDelega
...
+        epsiodeTableView.setShouldShowInfiniteScrollHandler { _ -> Bool in
+            return self.continueInfiniteScroll
+        }
         epsiodeTableView.register(EpisodeTableViewCell.self, forCellReuseIdentifier: "EpisodeTableViewCellIdentifier")
-        
         view.addSubview(epsiodeTableView)
+    
+        epsiodeTableView.infiniteScrollIndicatorView = createLoadingAnimationView()
+        
+        if let series = self.series {
```
Comment (dennisfedorko) [inline Podcast/SeriesDetailViewController.swift:61]: self.series should not be optional as that would render the entire VC useless, change the type of series to Series!

File: Podcast/EndpointRequest.swift (hunk 1 above)
Comment (natashaarmbrust) [reply Podcast/EndpointRequest.swift]: Needs to be changed because if there are no query parameters it should return nil ... rn it's returning [:] which gave me a bug in one of the endpoint requests

File: Podcast/EndpointRequest.swift (hunk 2 above)
Comment (natashaarmbrust) [reply Podcast/EndpointRequest.swift]: Again same reason as above

File: Podcast/Episode.swift (hunk 4 above)
Comment (natashaarmbrust) [reply Podcast/Episode.swift:85]: No because I was talking to joe and he plans to standardize the data more on his end and rn some use hh:mm:ss, mm:ss, hh:mm, etc .. this will be fixed later on once we get a standard format

File: Podcast/EpisodesBySeriesIdEndpointRequest.swift (hunk 7 above)
Comment (natashaarmbrust) [reply Podcast/EpisodesBySeriesIdEndpointRequest.swift]: The map gives you a tulle so you need the .1 .. also what would you prefer? A for loop?

File: Podcast/SearchTableViewController.swift (hunk 10 above)
Comment (natashaarmbrust) [reply Podcast/SearchTableViewController.swift]: Lol

File: Podcast/SeriesDetailHeaderView.swift (hunk 11 above)
Comment (natashaarmbrust) [reply Podcast/SeriesDetailHeaderView.swift]: It doesn't actually make one I was testing something out I'll delete that

File: Podcast/SeriesDetailViewController.swift (hunk 12 above)
Comment (natashaarmbrust) [reply Podcast/SeriesDetailViewController.swift:61]: Ok so I ran into a huge issue here with loading the table view with force unwrapping series .. the problem is we have series that can be passed in or we can make a network request for the series by the series id.. if we don't make it optional the view won't load properly

File: Podcast/SeriesDetailViewController.swift [hunk 13]
```diff
@@ -20,76 +21,139 @@ class SeriesDetailViewController: UIViewController, SeriesDetailHeaderViewDelega
...
-        s.episodes = [episode]
-        series = s
-    }
-    
-    func updateViewsWithSeries(series: Series) {
+    override func viewDidAppear(_ animated: Bool) {
+        super.viewDidAppear(animated)
+        if self.series != nil {
+            //load animation for awhile for better UX
+            let deadlineTime = DispatchTime.now() + .milliseconds(100)
```
Comment (dennisfedorko) [inline Podcast/SeriesDetailViewController.swift:79]: what is the point of this function?

File: Podcast/SeriesDetailViewController.swift [hunk 14]
```diff
@@ -20,76 +21,139 @@ class SeriesDetailViewController: UIViewController, SeriesDetailHeaderViewDelega
...
     }
     
-    func seriesDetailHeaderViewDidPressSubscribeButton(seriesDetailHeader: SeriesDetailHeaderView, subscribed: Bool) {
-        // Toggle subscribe (aka endpoint request)
-        
-        // When request comes back, update in case it failed
-//        seriesHeaderView.subscribeButtonChangeState(isSelected: subscribed)
+    //create and delete subscriptions
+    func seriesDetailHeaderViewDidPressSubscribeButton(seriesDetailHeader: SeriesDetailHeaderView) {
+        if !(series!.isSubscribed) { //subscribing to series
```
Comment (dennisfedorko) [inline Podcast/SeriesDetailViewController.swift]: don't need ()

Comment (dennisfedorko) [reply Podcast/SeriesDetailViewController.swift]: or comments

File: Podcast/SeriesDetailViewController.swift [hunk 15]
```diff
@@ -129,9 +198,9 @@ class SeriesDetailViewController: UIViewController, SeriesDetailHeaderViewDelega
         return sectionHeaderHeight
     }
     
-    func tableView(_ tableView: UITableView, willDisplayHeaderView view: UIView, forSection section: Int) {
+    func tableView(_ tableView: UITableView, viewForHeaderInSection section: Int) -> UIView? {
+        let view = UIView(frame: CGRect(x: 0, y: 0, width: self.view.frame.width, height: sectionHeaderHeight))
```
Comment (dennisfedorko) [inline Podcast/SeriesDetailViewController.swift]: rename headerView plz

File: Podcast/SeriesbySeriesIdEndpointRequest.swift [hunk 16]
```diff
@@ -0,0 +1,26 @@
...
+//  Podcast
+//
+//  Created by Natasha Armbrust on 3/30/17.
+//  Copyright © 2017 Cornell App Development. All rights reserved.
+//
+
+import UIKit
+import SwiftyJSON
+
+class SeriesBySeriesIdEndpointRequest: EndpointRequest {
```
Comment (dennisfedorko) [inline Podcast/SeriesbySeriesIdEndpointRequest.swift]: FetchSeriesForSeriesIDEndpointRequest

File: Podcast/TabbedPageViewController.swift [hunk 17]
```diff
@@ -251,9 +256,26 @@ class TabbedPageViewController: UIViewController, UIPageViewControllerDataSource
...
+        case .series:
+            guard let series = searchResults[.series]?[index] as? Series else { return }
+            searchResultsDelegate?.didTapOnSeriesCell(series: series)
+        case .people:
+            //present external profile view here
+            let dummyViewController = UIViewController()
+            presentingViewController?.navigationController?.pushViewController(dummyViewController, animated: true)
+        case .tags:
+            //present tag view here
+            let dummyViewController = UIViewController()
```
Comment (dennisfedorko) [inline Podcast/TabbedPageViewController.swift]: didn't you make the tag vc already?

File: Podcast/SeriesDetailViewController.swift (hunk 13 above)
Comment (natashaarmbrust) [reply Podcast/SeriesDetailViewController.swift:79]: oh okay so i thought UX perspective it looked a little weird to have the series just show up .. so now what happens is the loader animates for a second then series header view shows up. It looks better this way

File: Podcast/UserSubscriptionsEndpointRequest.swift [hunk 18]
```diff
@@ -0,0 +1,27 @@
...
+//  Podcast
+//
+//  Created by Natasha Armbrust on 3/29/17.
+//  Copyright © 2017 Cornell App Development. All rights reserved.
+//
+
+import UIKit
+import SwiftyJSON
+
+class UserSubscriptionsEndpointRequest: EndpointRequest {
```
Comment (dennisfedorko) [inline Podcast/UserSubscriptionsEndpointRequest.swift]: FetchUserSubscriptionsEndpointRequest

File: Podcast/TabbedPageViewController.swift (hunk 17 above)
Comment (natashaarmbrust) [reply Podcast/TabbedPageViewController.swift]: oh yea i can put that in

File: Podcast/UserSubscriptionsEndpointRequest.swift [hunk 19]
```diff
@@ -0,0 +1,27 @@
...
+    init(userID: String) {
+        self.userID = userID
+        super.init()
+        path = "/subscriptions"
+        httpMethod = .get
+        queryParameters = ["id": userID]
+    }
+    
+    override func processResponseJSON(_ json: JSON) {
+        processedResponseValue = json["data"]["subscriptions"].map{ series in SubscriptionSeries(json: series.1) }
```
Comment (dennisfedorko) [inline Podcast/UserSubscriptionsEndpointRequest.swift]: clean up map by making the type for series explicit

Comment (natashaarmbrust) [reply Podcast/UserSubscriptionsEndpointRequest.swift]: not quite sure what you mean by this

File: Podcast/EndpointRequest.swift (hunk 1 above)
Comment (dennisfedorko) [reply Podcast/EndpointRequest.swift]: what bug did it cause? this goes straight into the alamofire request initializer, so it shouldn't have any effect if its empty

File: Podcast/Episode.swift (hunk 4 above)
Comment (dennisfedorko) [reply Podcast/Episode.swift:85]: ok, mark with a //TODO

File: Podcast/EpisodesBySeriesIdEndpointRequest.swift (hunk 7 above)
Comment (dennisfedorko) [reply Podcast/EpisodesBySeriesIdEndpointRequest.swift]: why would mapping json give you a tuple? if you call some variable series, it should be assumed that its of type Series.

File: Podcast/SeriesDetailViewController.swift (hunk 12 above)
Comment (dennisfedorko) [reply Podcast/SeriesDetailViewController.swift:61]: then clearly something is wrong with the setup of the controller. in what case would we pass it in and in which case can we make a network request?

File: Podcast/SeriesDetailViewController.swift (hunk 13 above)
Comment (dennisfedorko) [reply Podcast/SeriesDetailViewController.swift:79]: ok ill take your word for it

File: Podcast/EndpointRequest.swift (hunk 1 above)
Comment (natashaarmbrust) [reply Podcast/EndpointRequest.swift]: It times out on the request cuz it's malformed

File: Podcast/EpisodesBySeriesIdEndpointRequest.swift (hunk 6 above)
Comment (natashaarmbrust) [reply Podcast/EpisodesBySeriesIdEndpointRequest.swift]: Addressed

File: Podcast/Episode.swift (hunk 3 above)
Comment (natashaarmbrust) [reply Podcast/Episode.swift:16]: Other PR included this

File: Podcast/Episode.swift (hunk 4 above)
Comment (natashaarmbrust) [reply Podcast/Episode.swift:85]: Addressed

File: Podcast/EpisodesBySeriesIdEndpointRequest.swift (hunk 5 above)
Comment (natashaarmbrust) [reply Podcast/EpisodesBySeriesIdEndpointRequest.swift]: Addressed

File: Podcast/SearchSeriesTableViewCell.swift (hunk 8 above)
Comment (natashaarmbrust) [reply Podcast/SearchSeriesTableViewCell.swift]: Addressed

File: Podcast/SeriesDetailViewController.swift (hunk 12 above)
Comment (natashaarmbrust) [reply Podcast/SeriesDetailViewController.swift:61]: Ok so in search we have full series we are passing in. Since we already get full series from search we don't need another network request before the view is presented but only need one to get the episodes. However in subscriptions since our data is denormalized on backend we only get part of a series so I make a network request to get the full series and show the loading indicator while I do that

File: Podcast/SeriesDetailViewController.swift (hunk 14 above)
Comment (natashaarmbrust) [reply Podcast/SeriesDetailViewController.swift]: Addressed

File: Podcast/SeriesDetailViewController.swift (hunk 15 above)
Comment (natashaarmbrust) [reply Podcast/SeriesDetailViewController.swift]: Addressed

File: Podcast/UserSubscriptionsEndpointRequest.swift (hunk 18 above)
Comment (natashaarmbrust) [reply Podcast/UserSubscriptionsEndpointRequest.swift]: Addressed

File: Podcast/EndpointRequest.swift (hunk 1 above)
Comment (natashaarmbrust) [reply Podcast/EndpointRequest.swift]: What happens is you call parameters() in the request which ends up being [:] if there are none. However alamofire is expecting nil if there are node

File: Podcast/EpisodesBySeriesIdEndpointRequest.swift (hunk 7 above)
Comment (natashaarmbrust) [reply Podcast/EpisodesBySeriesIdEndpointRequest.swift]: it returns (string, json) tuple indicating which number value it is in the json array ... i think its fine i'm not sure what else you'd like me to use

File: Podcast/SeriesDetailViewController.swift (hunk 12 above)
Comment (dennisfedorko) [reply Podcast/SeriesDetailViewController.swift:61]: I think we discussed this
